File: services/node/storage.js
'use strict'

// Imports
const helpers = require("./helpers.js")
const logger = require("./logger.js")
const mongodb = require("./mongodb.js")

const storage = {
    async editGodPage(req, res) {
        var perms = await helpers.authorize(req)
        switch (perms) {
            case "user", "admin":
                res.status(200).render("edit_god", {
                    title: "Edit God"
                })
                break

            case "no_auth":
                res.status(302).redirect("/login")
                break

            default:
                res.status(500).render("error", {
                    title: "Error",
                    httpCode: "500",
                    message: "There was an issue with the Server, please try again later."
                })
        }
    },

    async addGodPage(req, res) {
        var perms = await helpers.authorize(req)
        switch (perms) {
            case "user", "admin":
                res.status(200).render("add_god", {
                    title: "Add God"
                })
                break

            case "no_auth":
                res.status(302).redirect("/login")
                break

            default:
                res.status(500).render("error", {
                    title: "Error",
                    httpCode: "500",
                    message: "There was an issue with the Server, please try again later."
                })
        }
    },

    async storagePage(req, res) {
        var perms = await helpers.authorize(req)
        switch (perms) {
            case "user", "admin":
                res.status(200).render("storage", {
                    title: "God Storage"
                })
                break

            case "no_auth":
                res.status(302).redirect("/login")
                break

            default:
                res.status(500).render("error", {
                    title: "Error",
                    httpCode: "500",
                    message: "There was an issue with the Server, please try again later."
                })
        }
    },

    async storageSearchAPI(req, res) {
        var perms = await helpers.authorize(req)
        switch (perms) {
            case "user", "admin":
                if (!await helpers.validate(req.body, [["filter", "[a-zA-Z]{0,32}"]])) {
                    res.status(400).json({
                        message: "There was an issue with your request.",
                    })
                    return
                }

                var result = await mongodb.collection("gods").find({ name: new RegExp(req.body.filter) }).toArray()
                res.status(200).json({
                    message: "Successfully filtered gods.",
                    result:  result,
                })
                break

            case "no_auth":
                res.status(401).json({
                    message: "Your credentials are invalid."
                })
                break

            default:
                res.status(500).render({
                    message: "There was an issue with the Server, please try again later."
                })
        }
    }
}

module.exports = storage

File: services/golang/vesuvius/storage.go
package main

import (
	"net/http"

	"github.com/gin-gonic/gin"
	"go.mongodb.org/mongo-driver/bson"
	"go.mongodb.org/mongo-driver/bson/primitive"
	gintrace "gopkg.in/DataDog/dd-trace-go.v1/contrib/gin-gonic/gin"
)

func EditGodPage(ctx *gin.Context) {
	permissions := Authenticate(ctx)
	switch permissions {
	case "user", "admin":
		gintrace.HTML(ctx, http.StatusOK, "edit_god.html", gin.H{
			"title":    "Edit God",
			"language": "GO",
		})

	case "none":
		ctx.Redirect(http.StatusFound, "/login")

	default:
		gintrace.HTML(ctx, http.StatusFound, "error.html", gin.H{
			"title":    "Error",
			"language": "GO",
			"httpCode": "500",
			"message":  "There was an issue with the Server, please try again later.",
		})
	}
}

func AddGodPage(ctx *gin.Context) {
	permissions := Authenticate(ctx)
	switch permissions {
	case "user", "admin":
		gintrace.HTML(ctx, http.StatusOK, "add_god.html", gin.H{
			"title":    "Add God",
			"language": "GO",
		})

	case "none":
		ctx.Redirect(http.StatusFound, "/login")

	default:
		gintrace.HTML(ctx, http.StatusFound, "error.html", gin.H{
			"title":    "Error",
			"language": "GO",
			"httpCode": "500",
			"message":  "There was an issue with the Server, please try again later.",
		})
	}
}

func StoragePage(ctx *gin.Context) {
	permissions := Authenticate(ctx)
	switch permissions {
	case "user", "admin":
		gintrace.HTML(ctx, http.StatusOK, "storage.html", gin.H{
			"title":    "God Storage",
			"language": "GO",
		})

	case "none":
		ctx.Redirect(http.StatusFound, "/login")

	default:
		gintrace.HTML(ctx, http.StatusFound, "error.html", gin.H{
			"title":    "Error",
			"language": "GO",
			"httpCode": "500",
			"message":  "There was an issue with the Server, please try again later.",
		})
	}
}

func StorageSearchAPI(ctx *gin.Context) {
	permissions := Authenticate(ctx)
	switch permissions {
	case "user", "admin":
		filter := make(map[string]string)
		ctx.ShouldBind(&filter)

		if !Validate(ctx, filter, [][2]string{{"filter", "[a-zA-Z]{0,32}"}}) {
			ctx.JSON(http.StatusBadRequest, gin.H{
				"message": "There was an issue with your request.",
			})
			return
		}

		var result []bson.M
		cursor, err := mongodb.Collection("gods").Find(ctx.Request.Context(), bson.M{"name": bson.M{"$regex": primitive.Regex{Pattern: filter["filter"], Options: "i"}}})
		if err != nil {
			if err.Error() != "mongo: no documents in result" {
				Log(ctx).WithError(err).Error(ctx.Error(err).Error())
				ctx.JSON(http.StatusInternalServerError, gin.H{
					"message": "There was an issue with the Server, please try again later.",
				})
				return
			}
		}

		err = cursor.All(ctx, &result)
		if err != nil {
			Log(ctx).WithField("result", result).WithError(err).Error(ctx.Error(err).Error())
			ctx.JSON(http.StatusInternalServerError, gin.H{
				"message": "There was an issue with the Server, please try again later.",
			})
			return
		}

		ctx.JSON(http.StatusOK, gin.H{
			"message": "Successfully filtered gods.",
			"result":  result,
		})

	case "none":
		ctx.JSON(http.StatusUnauthorized, gin.H{
			"message": "Your credentials are invalid.",
		})

	default:
		ctx.JSON(http.StatusInternalServerError, gin.H{
			"message": "There was an issue with the Server, please try again later.",
		})
	}
}
